fancy_mlp.py

```diff
@@ -2,64 +2,78 @@
+                layer_one_inputs = inputs
 
-		# Predict just like in simple_mlp.py
-		layer_one_prediction = applySigmoid(np.dot(layer_one_inputs, layer_one_weights))
+                # Predict just like in simple_mlp.py
+                layer_one_prediction = applySigmoid(np.dot(layer_one_inputs, layer_one_weights))
 
-		# Take the prediction from layer one and forward proogate it to the second layer of weights for a final output.
-		layer_two_prediction = applySigmoid(np.dot(layer_one_prediction, layer_two_weights))
+                # Take the prediction from layer one and forward proogate it to the second layer of weights for a final output.
+                layer_two_prediction = applySigmoid(np.dot(layer_one_prediction, layer_two_weights))
 
-		# How much were we off by?
-		layer_two_error = keys - layer_two_prediction 
+                # How much were we off by?
+                layer_two_error = keys - layer_two_prediction 
 
-		# Change in error just like in simple_mlp.py
-		layer_two_change_in_error = layer_two_error * applySigmoid(layer_two_prediction, True)
+                # Change in error just like in simple_mlp.py
+                layer_two_change_in_error = layer_two_error * applySigmoid(layer_two_prediction, True)
 
-		# Figure out how wrong our output for layer_one was by seeing how wrong the layer_two_prediction was
-		layer_one_error = np.dot(layer_two_change_in_error, layer_two_weights.T)
+                # Figure out how wrong our output for layer_one was by seeing how wrong the layer_two_prediction was
+                layer_one_error = np.dot(layer_two_change_in_error, layer_two_weights.T)
 
-		# Just like in simple_mlp.py
-		layer_one_change_in_error = layer_one_error * applySigmoid(layer_one_error, True)
+                # Just like in simple_mlp.py
+                layer_one_change_in_error = layer_one_error * applySigmoid(layer_one_prediction, True)
 
-		# adjust your weights accoridngly.
-		layer_one_weights +=  np.dot(layer_one_prediction.T, layer_one_change_in_error)
-		layer_two_weights +=  np.dot(layer_two_prediction.T, layer_two_change_in_error)
+                if iter == 0:
+                    assert layer_one_prediction.shape[0] == 4
+                    assert layer_one_prediction.shape[1] == 4
+                    assert layer_two_prediction.shape[0] == 4
+                    assert layer_two_prediction.shape[1] == 1
+                    assert layer_one_weights.shape[0] == 3
+                    assert layer_one_weights.shape[1] == 4
+                    assert layer_two_weights.shape[0] == 4
+                    assert layer_two_weights.shape[1] == 1
 
-		if(iter == 0 or iter == 5000 or iter == 9999):
-			print_data(iter, inputs, keys, weights, prediction)
+                # adjust your weights accoridngly.
+                layer_one_weights +=  np.dot(inputs.T, layer_one_change_in_error)
+                layer_two_weights +=  np.dot(layer_one_prediction.T,
```

Comment:
Then it's inconsistent with the ``simple_mlp.py``.
In there you update weights by doing: ``weights += np.dot(inputs.T, change_in_error)``, which works fine.

Have you tried running your ``fancy_mlp`` implementation?

First off it throws an error since the shapes do not match (See PR commit message for details).
Now if I only change ``layer_one_weights += np.dot(layer_one_prediction.T, layer_one_change_in_error)`` and ``layer_two_weights += np.dot(layer_two_prediction.T, layer_two_change_in_error)`` it does not converge.
This is the training output after 40000 steps:
```python
Output After Training:
[[0.49949126]
 [0.49944283]
 [0.49888797]
 [0.50120373]]
```


This is why I also change ``layer_one_change_in_error``. Again, this is consistent with the ``simple_mlp.py`` implementation.
And there is the result after 40000 steps: